[GCC] Fix build after 300911@main

dpino opened the pull request:
#### 0e023c832420d3c293dc452aad1823d29403a78e
<pre>
[GCC] Fix build after 300911@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=299972">https://bugs.webkit.org/show_bug.cgi?id=299972</a>

Reviewed by Adrian Perez de Castro.

GCC12 and lower return build error in &apos;StyleLengthWrapper.h&apos;:

error: invalid use of non-static data member ‘WebCore::Style::PrimitiveNumeric&lt;WebCore::CSS::Percentage&lt;WebCore::CSS::Range{0.0, +Inf, WebCore::CSS::RangeClampOptions::Default, WebCore::CSS::RangeZoomOptions::Default}, float&gt; &gt;::value’

Changeset 298558@main fixed the same kind of error, although related to
compilation of PrimitiveKeyword in GCC12 or lower. The solution was to
fix the build failure by adding an alternative implementation of the
faulting code specific for GCC12 or lower.

This patch follows a similar approach.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::isSelfCollapsingBlock const):

Canonical link: <a href="https://commits.webkit.org/301265@main">https://commits.webkit.org/301265@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0944c98454dd09c1d99c781a8358dbca53fd8641

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125352 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45017 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35758 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132202 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77219 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127229 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45705 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53582 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95439 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63362 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128306 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36505 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112100 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75979 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35404 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30268 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75679 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106273 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30494 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134887 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52155 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39934 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103912 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52593 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108321 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103671 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49037 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27332 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49272 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19638 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52050 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57829 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51407 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54763 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53101 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d215e9c9c1b9b37e5c951afaa4985a06a78c8ee5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51934

dpino commented:
> So we have a theory about whether this is a GCC bug or feature? If bug did someone report it to the GCC project?

Definitely it's a bug in GCC since GCC13 builds fine. The WebKitGTK/WPEWebKit EWS bots use the FlatpakSDK to build, which features GCC13. That's why they do not report a build error on this code. The post-commit bots for Ubuntu or Debian build with the GCC version shipped in their OS, which in the case of Debian 12 is GCC12. This is the oldest version of GCC we're supporting now.

I will try to locate the GCC error in GCC's git repo.

dpino committed:
[GCC] Fix build after 300911@main
https://bugs.webkit.org/show_bug.cgi?id=299972

Reviewed by Adrian Perez de Castro.

GCC12 and lower return build error in 'StyleLengthWrapper.h':

error: invalid use of non-static data member ‘WebCore::Style::PrimitiveNumeric<WebCore::CSS::Percentage<WebCore::CSS::Range{0.0, +Inf, WebCore::CSS::RangeClampOptions::Default, WebCore::CSS::RangeZoomOptions::Default}, float> >::value’

Changeset 298558@main fixed the same kind of error, although related to
compilation of PrimitiveKeyword in GCC12 or lower. The solution was to
fix the build failure by adding an alternative implementation of the
faulting code specific for GCC12 or lower.

This patch follows a similar approach.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::isSelfCollapsingBlock const):

Canonical link: https://commits.webkit.org/301265@main
